src/task/mod.rs

```diff
@@ -36,30 +35,86 @@ pub trait Task: Send + Sized + 'static {
     /// function callback(err, value) {}
     /// ```
     fn schedule(self, callback: Handle<JsFunction>) {
-        let boxed_self = Box::new(self);
-        let self_raw = Box::into_raw(boxed_self);
-        let callback_raw = callback.to_raw();
-        unsafe {
-            neon_runtime::task::schedule(mem::transmute(self_raw),
-                                         perform_task::<Self>,
-                                         complete_task::<Self>,
-                                         callback_raw);
-        }
+        schedule(move || {
+            let result = self.perform();
+
+            move |cx| self.complete(cx, result)
+        }, callback);
+    }
+}
+
+pub struct TaskBuilder<'c, C, Perform> {
+    context: &'c mut C,
+    perform: Perform,
+}
+
+impl<'c, C, Perform> TaskBuilder<'c, C, Perform> {
+    pub(crate) fn new(context: &'c mut C, perform: Perform) -> Self {
+        TaskBuilder { context, perform }
+    }
+}
+
+impl<'a, 'c, C, Perform, Complete, Output> TaskBuilder<'c, C, Perform>
+where
+    C: Context<'a>,
+    Perform: FnOnce() -> Complete + Send + 'static,
+    Complete: FnOnce(TaskContext) -> JsResult<Output> + Send + 'static,
+    Output: Value,
+{
+    pub fn schedule(self, callback: Handle<JsFunction>) -> JsResult<'a, JsValue> {
+        let Self { perform, context } = self;
+
+        schedule(perform, callback);
+
+        Ok(context.undefined().upcast())
     }
 }
 
-unsafe extern "C" fn perform_task<T: Task>(task: *mut c_void) -> *mut c_void {
-    let task: Box<T> = Box::from_raw(mem::transmute(task));
-    let result = task.perform();
-    Box::into_raw(task);
-    mem::transmute(Box::into_raw(Box::new(result)))
+fn schedule<Perform, Complete, Output>(
+    perform: Perform,
+    callback: Handle<JsFunction>,
+)
+where
+    Perform: FnOnce() -> Complete + Send + 'static,
+    Complete: FnOnce(TaskContext) -> JsResult<Output> + Send + 'static,
+    Output: Value,
+{
+    let data = Box::into_raw(Box::new(perform));
+
+    unsafe {
+        neon_runtime::task::schedule(
+            data as *mut _,
```

Comment:
`transmute` is more powerful and _dangerous_ than casting. Instead, use casting on the pointer types.